src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,473 @@
+        }
+
+        // Create complete point addition gate
+        {
+            let add_config: add::Config = (&config).into();
+            add_config.create_gate(meta);
+        }
+
+        // Create fixed-base scalar mul gates
+        {
+            let mul_fixed_config: mul_fixed::Config = (&config).into();
+            mul_fixed_config.create_gate::<C>(meta);
+        }
+
+        // Create variable-base scalar mul gates
+        {
+            let mul_config: mul::Config<C> = (&config).into();
+            mul_config.create_gate(meta);
+        }
+
+        config
+    }
+
+    pub fn load() -> <Self as Chip<C::Base>>::Loaded {
+        let commit_ivk_r = load::commit_ivk_r();
+        let note_commit_r = load::note_commit_r();
+        let nullifier_k = load::nullifier_k();
+        let value_commit_r = load::value_commit_r();
+        let value_commit_v = load::value_commit_v();
+
+        EccLoaded {
+            commit_ivk_r,
+            note_commit_r,
+            nullifier_k,
+            value_commit_r,
+            value_commit_v,
+        }
+    }
+}
+
+/// A full-width scalar used for variable-base scalar multiplication.
+/// This is decomposed in chunks of `window_width` bits in little-endian order.
+/// For example, if `window_width` = 3, we will have [k_0, k_1, ..., k_n]
+/// where `scalar = k_0 + k_1 * (2^3) + ... + k_n * (2^3)^n`.
+#[derive(Clone, Debug)]
+pub struct EccScalarFixed<C: CurveAffine> {
+    value: Option<C::Scalar>,
+    k_bits: Vec<CellValue<C::Base>>,
+}
+
+/// A signed short scalar used for variable-base scalar multiplication.
+/// This is decomposed in chunks of `window_width` bits in little-endian order.
+/// For example, if `window_width` = 3, we will have [k_0, k_1, ..., k_n]
+/// where `scalar = k_0 + k_1 * (2^3) + ... + k_n * (2^3)^n`.
+#[derive(Clone, Debug)]
+pub struct EccScalarFixedShort<C: CurveAffine> {
+    magnitude: Option<C::Scalar>,
+    sign: CellValue<C::Base>,
+    k_bits: Vec<CellValue<C::Base>>,
+}
+
+impl<C: CurveAffine> EccInstructions<C> for EccChip<C> {
+    type ScalarFixed = EccScalarFixed<C>;
+    type ScalarFixedShort = EccScalarFixedShort<C>;
+    type ScalarVar = CellValue<C::Base>;
+    type Point = EccPoint<C>;
+    type X = CellValue<C::Base>;
+    type FixedPoint = OrchardFixedBase<C>;
+    type FixedPointShort = OrchardFixedBaseShort<C>;
+    type FixedPoints = OrchardFixedBases<C>;
+    type FixedPointsShort = OrchardFixedBasesShort<C>;
+
+    fn witness_scalar_var(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Base>,
+    ) -> Result<Self::ScalarVar, Error> {
+        layouter.assign_region(
+            || "Witness scalar var",
+            |mut region| {
+                let cell = region.assign_advice(
+                    || "Scalar var",
+                    self.config().P.0,
+                    0,
+                    || value.ok_or(Error::SynthesisError),
+                )?;
```

Comment:
I'm still confused about this.

We're witnessing a scalar of Pallas (an element of Fq) in our circuit which is built over Fp (the native circuit field and the scalar field for Vesta, the curve we're using for halo in orchard). But q < p! So some scalars (Fq) cannot fit inside of a single cell in Fp. What am I missing?

Reply:
This scalar multiplication only needs to work for scalars in [0, p).